service/oauth2/data_plane.go

```diff
@@ -0,0 +1,50 @@
+package oauth2
+
+import (
+	"strings"
+
+	"golang.org/x/oauth2"
+)
+
+// DataPlaneHelper is a helper struct to fetch, refresh and manage DataPlane details and tokens.
+type DataPlaneHelper struct {
+	infos  map[string]*DataPlaneInfo
+	tokens map[string]*oauth2.Token
+}
+
+// GetDataPlaneDetails returns the endpoint URL and token. It returns a cached token if it is valid,
+// otherwise it refreshes the token and returns the new token.
+func (o *DataPlaneHelper) GetDataPlaneDetails(method string, params []string, refresh func(*DataPlaneInfo) (*oauth2.Token, error), infoGetter func() (*DataPlaneInfo, error)) (string, *oauth2.Token, error) {
+	if o.infos == nil {
```

Comment:
I think we need a mutex to protect access to these fields, since they are stateful and can be called concurrently. Might be a good candidate for a double check, since this is only done once the first time a dataplane API is called for a service:
```go
if o.infos == nil {
  lock()
  if o.infos == nil {
    initialize
  }
  unlock()
}
```

Reply:
Done for the ones below. For the maps, I created a constructor method.